src/nvim/terminal.c

```diff
@@ -907,6 +928,16 @@ static void on_refresh(Event event)
       }
       continue;
     }
+    FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+      if (wp->w_buffer == term->buf) {
+        if ( wp->w_botline > wp->w_buffer->b_ml.ml_line_count ) {
+          // if last line is visible, follow the terminal
+          wp->following_terminal = true;
+        } else {
+          wp->following_terminal = false;
+        }
```

Comment:
I've been thinking about it some more, and I'd propose the following:
* terminal view will scroll whenever vim's cursor is at the same line or below where the terminal thinks the cursor is - i.e. usually the last non-blank line (in normal terminal/bash operation), although not always the actual last line of the buffer. Position within that line won't matter.
* scrollback should always have at least one line

This allows the following scenarios :
* you're in terminal mode, pretty much anywhere on screen (just opened a terminal, just cleared it, or way at the bottom of a command with lots of output). You go to normal mode, perhaps switch to another buffer or whatever. Terminal spews output and the screen follows.
* Same scenario, but you go to normal mode and up one line. Screen stays where it is. 
* You go to a terminal screen. You press G and you're on the bottom and following.

All three are important scenarios I think.

If the position within the line would matter, you couldn't go to follow mode as easily. If Vim's cursor would have to be at the actual bottom (blank) line to follow, and you've just cleared the terminal, entered a command and gone back to normal mode, it wouldn't follow (in the new behavior where Vim's cursor picks up where the terminal's left off).

I think/hope this would also meet your expectations, if not, please let me know.

I'll report back when I've found the time to implement this - not sure yet when :-)